core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -521,16 +509,17 @@ private[spark] class BlockManager(
               }
             }
           }
+        } else {
+          releaseLock(blockId)
```

Comment:
It took me a while to figure out what this is case is. IIUC this happens when there is an entry for this block in `BlockInfoManager`, but it is neither in memory nor marked to use disk. In other words, this is only for `MEMORY_ONLY` and after the block has already been evicted. In that case it would make sense to release the lock as you have done.

I'm actually not sure if we'll ever hit this case (if the block has already been evicted then its metadata won't even be in `BlockInfoManager`). Either way a simple comment to explain what this case represents will be very helpful.